src/ring_theory/polynomial/chebyshev.lean

```diff
@@ -0,0 +1,301 @@
+
+/-!
+# Chebyshev polynomials
+
+The Chebyshev polynomials are two families of polynomials indexed by `ℕ`, with integral coefficients.
+In this file, we only consider Chebyshev polynomials of the first kind.
+
+## Main declarations
+
+* `polynomial.chebyshev₁`: the Chebyshev polynomials of the first kind.
+* `polynomial.lambdashev`: a variant on the Chebyshev polynomials that define a Lambda structure
+  on `polynomial ℤ`.
+* `polynomial.chebyshev₁_complex_cos`: the `n`-th Chebyshev polynomial evaluates on `complex.cos θ`
+  to the value `n * complex.cos θ`.
+* `polynomial.chebyshev₁_mul`, the `(m * n)`-th Chebyshev polynomial is the composition
+  of the `m`-th and `n`-th Chebyshev polynomials.
+* `polynomial.lambdashev_mul`, the `(m * n)`-th lambdashev polynomial is the composition
+  of the `m`-th and `n`-th lambdashev polynomials.
+* `polynomial.lambdashev_char_p`, for a prime number `p`, the `p`-th lambdashev polynomial
+  is congruent to `X ^ p` modulo `p`.
+
+## Implementation details
+
+Since Chebyshev polynomials have interesting behaviour over the complex numbers and modulo `p`,
+we define them to have coefficients in an arbitrary commutative ring, even though
+technically `ℤ` would suffice.
+The benefit of allowing arbitrary coefficient rings, is that the statements afterwards are clean,
+and do not have `map (int.cast_ring_hom R)` interfering all the time.
+
+## TODO
+
+Add Chebyshev polynomials of the second kind.
+
+-/
+
+noncomputable theory
+
+namespace polynomial
+
+variables (R S : Type*) [comm_ring R] [comm_ring S]
+
+/-- `chebyshev₁ n` is the `n`-th Chebyshev polynomial of the first kind -/
+noncomputable def chebyshev₁ : ℕ → polynomial R
+| 0       := 1
+| 1       := X
+| (n + 2) := 2 * X * chebyshev₁ (n + 1) - chebyshev₁ n
+
+@[simp] lemma chebyshev₁_zero : chebyshev₁ R 0 = 1 := rfl
+@[simp] lemma chebyshev₁_one : chebyshev₁ R 1 = X := rfl
+lemma chebyshev₁_two : chebyshev₁ R 2 = 2 * X ^ 2 - 1 :=
+by simp only [chebyshev₁, sub_left_inj, pow_two, mul_assoc]
+@[simp] lemma chebyshev₁_add_two (n : ℕ) :
+  chebyshev₁ R (n + 2) = 2 * X * chebyshev₁ R (n + 1) - chebyshev₁ R n :=
+by rw chebyshev₁
+
+lemma chebyshev₁_two_le (n : ℕ) (h : 2 ≤ n) :
+  chebyshev₁ R n = 2 * X * chebyshev₁ R (n - 1) - chebyshev₁ R (n - 2) :=
+begin
+  obtain ⟨n, rfl⟩ := nat.exists_eq_add_of_le h,
+  rw add_comm,
+  exact chebyshev₁_add_two R n
+end
+
+variables {R S}
+
+lemma map_chebyshev₁ (f : R →+* S) :
+  ∀ (n : ℕ), map f (chebyshev₁ R n) = chebyshev₁ S n
+| 0       := by simp only [chebyshev₁_zero, map_one]
+| 1       := by simp only [chebyshev₁_one, map_X]
+| (n + 2) :=
+begin
+  simp only [chebyshev₁_add_two, map_mul, map_sub, map_X, bit0, map_add, map_one],
+  rw [map_chebyshev₁ (n + 1), map_chebyshev₁ n],
+end
+
+variables (R)
+
+section complex
+open complex
+
+/-- the `n`-th Chebyshev polynomial evaluates on `cos θ` to the value `n * cos θ` -/
+@[simp] lemma chebyshev₁_complex_cos (θ : ℂ) :
+  ∀ n, (chebyshev₁ ℂ n).eval (cos θ) = cos (n * θ)
```

Comment:
One could argue that the `simp` should go in the other direction!  (I think @PatrickMassot was thinking along these lines.). That is, that the relevant simp-ing of Chebyshev polynomials is to let you reduce the number of different arguments fed to `cos` within a single expression.  Is there a reason you chose this direction for your simp?

Reply:
Also, perhaps we could move this `complex` section to `analysis.special_functions.trigonometric`?  It'd be nice to be able to refer to the Chebyshev polynomials there.

Reply:
I've done this.